explainer-beforematch.md

```diff
@@ -0,0 +1,135 @@
+page which allows custom styling of the matched element, which is now only
+possible with approximations from scroll positions and intersection
+observations.
+
+### Use Cases
+
+#### Example 1
+
+```html
+<style>
+.section {
+  transition: background 2s;
+}
+.highlighted {
+  background: cornsilk;
+  transition: background 0s;
+}
+</style>
+
+<div class=section>Lorem ipsum dolor sit amet</div>
+<div class=section>consectetur adipiscing elit</div>
+<div class=section>Sed augue lacus</div>
+
+<script>
+function highlight(e) {
+  e.target.classList.add("highlighted");
+  requestAnimationFrame(
+    () => requestAnimationFrame(
+      () => e.target.classList.remove("highlighted")));
+}
+
+document.querySelectorAll(".section").forEach(item => {
+  item.addEventListener("beforematch", highlight);
+});
+</script>
+```
+
+In this example, a highlighted entry gains a corksilk background color, which
+fades over approximately two seconds. It's a subtle effect that grabs the user's
+attention but is not too intrusive. 
+
+We use a `beforematch` event to highlight the section. Note that the highlighted
+element is the target element of the `beforematch` event, but we can also modify
+any related style or DOM based on the target's location.
+
+The effect of this example is that the section is briefly highlighted in
+cornsilk on an active find-in-page match. This effect happens in addition to the
+user-agent highlight the found match.
+
+#### Example 2
+```html
+<style>
+.collapsed {
+  subtree-visibility: hidden-matchable;
+}
+</style>
+
+Please explore the following sections:
+<h1>Introduction</h1> <div class=collapsed>...</div>
+<h1>Thesis</h1>       <div class=collapsed>...</div>
+<h1>Conclusion</h1>   <div class=collapsed>...</div>
+
+<script>
+function expand(e) {
+  e.target.classList.remove("collapsed");
+}
+
+document.querySelectorAll(".collapsed").forEach(item => {
+  item.addEventListener("beforematch", expand);
+});
+</script>
+```
+
+In this example, most of the content of the page is hidden in collapsed section.
+It uses a display-locking property called `subtree-visibility` to hide the
+content while letting it be searchable by find-in-page. When a match is found,
+and `beforematch` event is fired, we expand the section by removing the
+collapsed class.
+
+Note that the net effect of this is that the user is able to find matches in
+collapsed sections which are automatically expanded. This is possible due to
+both display-locking functionality and the `beforematch` event.
+
+Also note that the adoption of the `beforematch` in these use-cases should be
+straight-forward, since the typical page that provides content in collapsed
+sections already has an event handler to expand and show the section. With the
+`beforematch` event, we can reuse the same handler to expand the section.
+
+### Privacy Concerns
+
+Note that this event exposes more information to the origin that would otherwise
```

Comment:
Should "origin" be "page" instead to make the use of vocabulary more consistent? Or is there a distinction?